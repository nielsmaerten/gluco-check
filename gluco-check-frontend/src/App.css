@import url("https://fonts.googleapis.com/css2?family=Poppins&display=swap");
@import url("https://fonts.googleapis.com/icon?family=Material+Icons");
