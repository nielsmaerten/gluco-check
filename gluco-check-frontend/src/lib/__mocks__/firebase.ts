import { BloodGlucoseUnit, DiabetesMetric } from "../enums";
import { GlucoCheckUserDocument } from "../types";

export const mockUser: Partial<firebase.User> = {
  email: "example@example.com",
  uid: "uid123",
};

export const mockUserDocument: GlucoCheckUserDocument = {
  heardDisclaimer: false,
  defaultMetrics: [DiabetesMetric.BloodSugar],
  glucoseUnit: BloodGlucoseUnit.mgdl,
  nightscout: {
    url: "https://example.com",
    token: "nstoken123",
  },
};
