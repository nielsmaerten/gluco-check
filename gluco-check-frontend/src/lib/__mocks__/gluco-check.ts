import { plainToClass } from "class-transformer";
import { NightscoutValidationEndpointResponseDto } from "../NightscoutValidationClient/NightscoutValidationClientDto";

export const MOCK_NSV_RESPONSE_VALID = {
  token: {
    isValid: true,
    parsed: "token123",
  },
  url: {
    isValid: true,
    parsed: "https://example.com",
    pointsToNightscout: true,
  },
  nightscout: {
    glucoseUnit: "mmol",
    minSupportedVersion: "13.0.0",
    version: "14.0.7",
  },
  discoveredMetrics: [
    "blood sugar",
    "insulin on board",
    "carbs on board",
    "sensor age",
    "cannula age",
    "pump battery",
  ],
};

export const MOCK_NSV_RESPONSE_NON_NS_URL = {
  url: {
    parsed: "https://example.com",
    isValid: true,
    pointsToNightscout: false,
  },
  token: {
    parsed: "",
    isValid: false,
  },
  nightscout: {
    version: "",
    glucoseUnit: "",
    minSupportedVersion: "13.0.0",
  },
  discoveredMetrics: [],
};

export const MOCK_NSV_RESPONSE_NS_NEEDS_UPGRADE = {
  token: {
    isValid: false,
    parsed: "token123",
  },
  url: {
    isValid: true,
    parsed: "https://example.com",
    pointsToNightscout: true,
  },
  nightscout: {
    glucoseUnit: "mmol",
    minSupportedVersion: "13.0.0",
    version: "12.9.9",
  },
  discoveredMetrics: [],
};

export const MOCK_NSV_RESPONSE_INVALID_TOKEN = {
  url: {
    parsed: "https://example.com",
    isValid: true,
    pointsToNightscout: true,
  },
  token: {
    isValid: false,
    parsed: "token1234",
  },
  nightscout: {
    minSupportedVersion: "13.0.0",
    glucoseUnit: "mmol",
    version: "14.0.7",
  },
  discoveredMetrics: ["blood sugar"],
};

export const mockNsvResponseDtoValid = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_VALID
);

export const mockNsvResponseDtoNonNsUrl = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_NON_NS_URL
);

export const mockNsvResponseDtoNsNeedsUpgrade = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_NS_NEEDS_UPGRADE
);

export const mockNsvResponseDtoInvalidToken = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_INVALID_TOKEN
);
