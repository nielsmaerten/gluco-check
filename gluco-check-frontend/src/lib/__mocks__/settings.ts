import { BloodGlucoseUnit, DiabetesMetric } from "../enums";
import { SettingsFormData } from "../types";

export const mockFormData: SettingsFormData = {
  defaultMetrics: [DiabetesMetric.BloodSugar],
  nightscoutToken: "I am a token",
  nightscoutUrl: "https://example.com",
  glucoseUnit: BloodGlucoseUnit.mgdl,
};
