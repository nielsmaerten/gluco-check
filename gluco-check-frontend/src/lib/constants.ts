import { BloodGlucoseUnit } from "./enums";

export const FIRESTORE_FIELD_PATH_GLUCOSE_UNITS = "glucoseUnit";
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT = "nightscout";
export const FIRESTORE_FIELD_PATH_DEFAULT_METRICS = "defaultMetrics";
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT_URL = `${FIRESTORE_FIELD_PATH_NIGHTSCOUT}.url`;
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT_TOKEN = `${FIRESTORE_FIELD_PATH_NIGHTSCOUT}.token`;
export const FIRESTORE_FIELD_HEARD_DISCLAIMER = "heardDisclaimer";

export const NIGHTSCOUT_VALIDATION_ENDPOINT_URL =
  process.env.REACT_APP_NIGHTSCOUT_VALIDATION_ENDPOINT_URL || "";

export const DEFAULT_GLUCOSE_UNITS = BloodGlucoseUnit.mmol;

export const ALERT_AUTOHIDE_DURATION = 7000;
export const VALIDATION_DEBOUNCE_DURATION = 500;
export const DEFAULT_ONBOARDING_CAROUSEL_INTERVAL = 5000;

export const APP_DEBUG = process.env.REACT_APP_DEBUG === "true" || false;
