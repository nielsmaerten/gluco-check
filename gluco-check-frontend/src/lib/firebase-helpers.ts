import { BloodGlucoseUnit, DiabetesMetric } from "./enums";
import { GlucoCheckUserDocument } from "./types";

export const FIRESTORE_DEFAULT_SET_OPTIONS: firebase.firestore.SetOptions = {
  mergeFields: [
    "defaultMetrics",
    "glucoseUnit",
    "nightscout.token",
    "nightscout.url",
  ],
};

export const DEFAULT_USER_DOCUMENT: GlucoCheckUserDocument = {
  defaultMetrics: [DiabetesMetric.BloodSugar],
  glucoseUnit: BloodGlucoseUnit.mgdl,
  nightscout: {
    token: "",
    url: "",
  },
};

export const getDocumentPathForUser = (user: firebase.User): string => {
  return `users/${user.email}`;
};
