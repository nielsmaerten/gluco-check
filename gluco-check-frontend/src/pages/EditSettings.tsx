import React, { useContext } from "react";
import { useDocument } from "react-firebase-hooks/firestore";
import { firestore } from "../lib/firebase";
import {
  DEFAULT_USER_DOCUMENT,
  FIRESTORE_DEFAULT_SET_OPTIONS,
} from "../lib/firebase-helpers";
import { Alert } from "@material-ui/lab";
import { Container, makeStyles, Typography } from "@material-ui/core";
import { useTranslation } from "react-i18next";

import SettingsForm from "../components/SettingsForm";
import { SettingsFormData } from "../lib/types";
import {
  APP_DEBUG,
  DEFAULT_GLUCOSE_UNITS,
  FIRESTORE_FIELD_HEARD_DISCLAIMER,
  FIRESTORE_FIELD_PATH_DEFAULT_METRICS,
  FIRESTORE_FIELD_PATH_GLUCOSE_UNITS,
  FIRESTORE_FIELD_PATH_NIGHTSCOUT_TOKEN,
  FIRESTORE_FIELD_PATH_NIGHTSCOUT_URL,
  NIGHTSCOUT_VALIDATION_ENDPOINT_URL,
} from "../lib/constants";
import { userSettingsFormDataToUserSettingsDocument } from "../lib/transform";
import { FirebaseUserDocumentContext } from "../App";
import { NightscoutValidationClient } from "../lib/NightscoutValidationClient/NightscoutValidationClient";

export const returnHandleSettingsSave = (userDocumentPath: string) => {
  return async (data: SettingsFormData) => {
    const formDataAsDocument = userSettingsFormDataToUserSettingsDocument(data);
    try {
      await firestore.doc(userDocumentPath).set(
        {
          ...DEFAULT_USER_DOCUMENT,
          ...formDataAsDocument,
        },
        FIRESTORE_DEFAULT_SET_OPTIONS
      );
    } catch (e) {
      throw new Error(`error saving document ${e.message}`);
    }
  };
};

const useStyles = makeStyles((theme) => ({
  container: {
    paddingTop: theme.spacing(2),
    "& h2": {
      marginBottom: theme.spacing(2),
    },
  },
  alert: {
    marginBottom: theme.spacing(2),
  },
}));

export default function EditSettings() {
  const classes = useStyles();
  const userDocumentPath = useContext(FirebaseUserDocumentContext);
  const userDocumentReference = firestore.doc(userDocumentPath);

  const [document, loading, error] = useDocument(userDocumentReference);

  const { t } = useTranslation();

  const nightscoutUrl =
    document?.get(FIRESTORE_FIELD_PATH_NIGHTSCOUT_URL) ?? "";
  const nightscoutToken =
    document?.get(FIRESTORE_FIELD_PATH_NIGHTSCOUT_TOKEN) ?? "";
  const defaultMetrics =
    document?.get(FIRESTORE_FIELD_PATH_DEFAULT_METRICS) ?? [];
  const glucoseUnit =
    document?.get(FIRESTORE_FIELD_PATH_GLUCOSE_UNITS) ?? DEFAULT_GLUCOSE_UNITS;
  const hasHeardDisclaimer =
    document?.get(FIRESTORE_FIELD_HEARD_DISCLAIMER) ?? false;

  return (
    <Container maxWidth="md" className={classes.container}>
      <Typography variant="h6" component="h2">
        {t("settings.title")}
      </Typography>
      {!hasHeardDisclaimer && (
        <Alert severity="warning" className={classes.alert}>
          {t("settings.betaBanner")}
        </Alert>
      )}

      {loading && <>{t("status.general.loading")}</>}
      {error && (
        <>
          {t("status.general.error")}: {error.message}
        </>
      )}
      {document && (
        <SettingsForm
          nightscoutUrl={nightscoutUrl}
          nightscoutToken={nightscoutToken}
          glucoseUnit={glucoseUnit}
          defaultMetrics={defaultMetrics}
          onSubmit={returnHandleSettingsSave(userDocumentPath)}
          nightscoutValidator={
            new NightscoutValidationClient({
              endpointUrl: NIGHTSCOUT_VALIDATION_ENDPOINT_URL,
            })
          }
          shouldShowGlucoseUnitsField={APP_DEBUG}
        />
      )}
    </Container>
  );
}
