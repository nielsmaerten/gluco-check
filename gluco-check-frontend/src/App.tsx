import React from "react";
import { auth } from "./lib/firebase";
import { useAuthState } from "react-firebase-hooks/auth";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link,
  Redirect,
} from "react-router-dom";
import { getDocumentPathForUser } from "./lib/firebase-helpers";
import { useTranslation } from "react-i18next";
import {
  AppBar,
  Container,
  IconButton,
  makeStyles,
  Paper,
  Toolbar,
  Typography,
} from "@material-ui/core";
import { GitHub, Help } from "@material-ui/icons";

import Landing from "./pages/Landing";
import LanguageSelector from "./components/LanguageSelector";
import EditSettings from "./pages/EditSettings";
import Welcome from "./pages/Welcome";
import "./App.css";
import { FAQS_URL, GLUCO_CHECK_GITHUB_URL } from "./lib/constants";
import { AvailableLanguage } from "./lib/enums";

export const FirebaseUserDocumentContext = React.createContext("");

const useStyles = makeStyles((theme) => ({
  root: {
    "@media (prefers-reduced-motion: reduce)": {
      "& *": {
        animationDuration: "0.001ms !important",
        animationIterationCount: "1 !important",
        transitionDuration: "0.001ms !important",
      },
    },
  },
  container: {
    height: "100%",
  },
  surface: {
    padding: theme.spacing(2),
    marginBottom: theme.spacing(2),
    marginTop: theme.spacing(2),
  },
  leftToolbar: {},
  rightToolbar: {
    marginLeft: "auto", // switch to logical
    marginRight: -12,
  },
  nav: {
    "&, & > li": {
      margin: 0,
      padding: 0,
      listStyle: "none",
    },
    "& > li": {
      display: "inline-block",
    },
  },
  toolbar: {},
  navTitle: {
    "& a": {
      color: "white",
      textDecoration: "none",
    },
  },
}));

export default function App() {
  const classes = useStyles();
  const [user, loading] = useAuthState(auth);
  const { t, i18n } = useTranslation();
  const docPath = user ? getDocumentPathForUser(user) : ""; // TODO: we don't want this empty path to be possible, AND we want to auth protect any other authed routes

  let Content: React.ReactElement | null = null;
  if (user) {
    Content = <Welcome />;
  } else {
    if (!loading) {
      Content = <Landing />;
    }
  }

  const navigation = (
    <AppBar position="sticky">
      <Toolbar variant="regular" className={classes.toolbar}>
        <section className={classes.leftToolbar}>
          <Typography variant="h6" component="h1" className={classes.navTitle}>
            <Link to={`/${i18n.language}`}>{t("title")}</Link>
          </Typography>
        </section>
        <section className={classes.rightToolbar}>
          <ul className={classes.nav}>
            <li>
              <LanguageSelector />
            </li>
            <li>
              <IconButton
                aria-label={t("navigation.faqs")}
                color="inherit"
                data-testid="navigation-home"
                href={FAQS_URL}
              >
                <Help />
              </IconButton>
            </li>
            <li>
              <IconButton
                aria-label={t("navigation.github")}
                color="inherit"
                data-testid="navigation-contribute"
                href={GLUCO_CHECK_GITHUB_URL}
                target="_contribute"
              >
                <GitHub />
              </IconButton>
            </li>
          </ul>
        </section>
      </Toolbar>
    </AppBar>
  );

  const availableLanguageParams = Object.values(AvailableLanguage).join("|");
  const localesParamString = `/:locale(${availableLanguageParams})`;

  return (
    <div className={classes.root}>
      <Router>
        {navigation}
        <Container maxWidth="md" className={classes.container}>
          <Paper variant="elevation" className={classes.surface}>
            <Switch>
              <Redirect
                exact
                from="/settings"
                to={`/${i18n.language}/settings`}
              />
              <Redirect exact from="/" to={`/${i18n.language}`} />
              {user && (
                <Route path={`${localesParamString}/settings`}>
                  <FirebaseUserDocumentContext.Provider value={docPath}>
                    <EditSettings />
                  </FirebaseUserDocumentContext.Provider>
                </Route>
              )}
              <Route path={`${localesParamString}/`}>{Content}</Route>
            </Switch>
          </Paper>
        </Container>
      </Router>
    </div>
  );
}
