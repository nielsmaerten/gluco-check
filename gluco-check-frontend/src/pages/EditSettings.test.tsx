import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import * as firestoreHooks from "react-firebase-hooks/firestore";
import EditSettings, { returnHandleSettingsSave } from "./EditSettings";
import { FirebaseUserDocumentContext } from "../App";
import { firestore } from "../lib/firebase";
import { mockUserDocument } from "../lib/__mocks__/firebase";
import { mockFormData } from "../lib/__mocks__/settings";
import { FIRESTORE_DEFAULT_SET_OPTIONS } from "../lib/firebase-helpers";

// move to __mocks__
jest.mock("../lib/firebase.ts", () => {
  return {
    firestore: {
      doc: jest.fn().mockReturnValue({
        set: jest.fn(),
      }),
    },
  };
});

jest.mock("react-i18next", () => ({
  useTranslation: () => {
    return {
      t: jest.fn().mockImplementation((i) => {
        return i;
      }),
    };
  },
}));

jest.mock("../components/SettingsForm.tsx", () => {
  return {
    __esModule: true,
    default: () => {
      return <div>SettingsForm</div>;
    },
  };
});

afterEach(() => {
  cleanup();
  jest.clearAllMocks();
});

const mockDocPath = "users/alf@melmac.com";

describe("EditSettings", () => {
  const mockFirestoreDoc = {
    ...mockUserDocument,
    get: jest.fn().mockReturnValue("a value"),
  };
  it("Renders the component when loaded, with no axe violations", async () => {
    // @ts-ignore
    firestoreHooks.useDocument = jest
      .fn()
      .mockReturnValue([mockFirestoreDoc, false, undefined]);

    const { container } = render(
      <FirebaseUserDocumentContext.Provider value={mockDocPath}>
        <EditSettings />
      </FirebaseUserDocumentContext.Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });

  it("Renders the component when loaded and user has heard disclaimer, with no axe violations", async () => {
    // @ts-ignore
    firestoreHooks.useDocument = jest.fn().mockReturnValue([
      {
        ...mockFirestoreDoc,
        heardDiscalimer: true,
      },
      false,
      undefined,
    ]);

    const { container } = render(
      <FirebaseUserDocumentContext.Provider value={mockDocPath}>
        <EditSettings />
      </FirebaseUserDocumentContext.Provider>
    );

    expect(await axe(container)).toHaveNoViolations();
  });

  it("Renders the component when loading, with no axe violations", async () => {
    // @ts-ignore
    firestoreHooks.useDocument = jest
      .fn()
      .mockReturnValue([undefined, true, undefined]);

    const { container } = render(
      <FirebaseUserDocumentContext.Provider value={mockDocPath}>
        <EditSettings />
      </FirebaseUserDocumentContext.Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });

  it("Renders the component with error, with no axe violations", async () => {
    // @ts-ignore
    firestoreHooks.useDocument = jest
      .fn()
      .mockReturnValue([undefined, false, new Error("you blew it")]);

    const { container } = render(
      <FirebaseUserDocumentContext.Provider value={mockDocPath}>
        <EditSettings />
      </FirebaseUserDocumentContext.Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("returnHandleSettingsSave", () => {
  it("returns a callable submit handler", () => {
    const mockSet = jest.fn();
    (firestore.doc as jest.Mock).mockReturnValue({
      set: mockSet,
    });

    const handler = returnHandleSettingsSave(mockDocPath);
    expect(handler).toBeInstanceOf(Function);

    handler(mockFormData);
    expect(firestore.doc).toHaveBeenCalledWith(mockDocPath);
    expect(mockSet).toHaveBeenCalledTimes(1);
    expect(mockSet).toHaveBeenCalledWith(
      expect.any(Object),
      FIRESTORE_DEFAULT_SET_OPTIONS
    );
  });

  it("returns a callable submit handler that handles error", async () => {
    expect.assertions(1);
    const handler = returnHandleSettingsSave(mockDocPath);
    (firestore.doc as jest.Mock).mockReturnValue({
      set: jest.fn().mockImplementationOnce(() => {
        throw new Error("tried my best!");
      }),
    });
    try {
      await handler(mockFormData);
    } catch (e) {
      expect(e).toMatchInlineSnapshot(
        `[Error: error saving document tried my best!]`
      );
    }
  });
});
