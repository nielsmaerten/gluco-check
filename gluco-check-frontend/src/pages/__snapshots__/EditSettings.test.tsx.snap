// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSettings Renders the component when loaded, with no axe violations 1`] = `
<div
  class="MuiContainer-root makeStyles-container-1 MuiContainer-maxWidthMd"
>
  <h2
    class="MuiTypography-root MuiTypography-h6"
  >
    settings.title
  </h2>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardWarning makeStyles-alert-2 MuiPaper-elevation0"
    role="alert"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      settings.betaBanner
    </div>
  </div>
  <div>
    SettingsForm
  </div>
</div>
`;

exports[`EditSettings Renders the component when loading, with no axe violations 1`] = `
<div
  class="MuiContainer-root makeStyles-container-5 MuiContainer-maxWidthMd"
>
  <h2
    class="MuiTypography-root MuiTypography-h6"
  >
    settings.title
  </h2>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardWarning makeStyles-alert-6 MuiPaper-elevation0"
    role="alert"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      settings.betaBanner
    </div>
  </div>
  status.general.loading
</div>
`;

exports[`EditSettings Renders the component with error, with no axe violations 1`] = `
<div
  class="MuiContainer-root makeStyles-container-7 MuiContainer-maxWidthMd"
>
  <h2
    class="MuiTypography-root MuiTypography-h6"
  >
    settings.title
  </h2>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardWarning makeStyles-alert-8 MuiPaper-elevation0"
    role="alert"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      settings.betaBanner
    </div>
  </div>
  status.general.error
  : 
  you blew it
</div>
`;
